### PR TITLE
feat: upgrade React to v18

### PR DESCRIPTION
**Pull request by cjthompson**
## Description

**feat: upgrade React to v18**

* upgrade Chakra-UI to v2
* upgrade various react packages to latest versions where needed for compatibility

## Notice

- [X] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [X] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes https://github.com/shapeshift/web/issues/1739

## Risk

Maximum

## Testing

Everything

## Screenshots (if applicable)

N/A

### CODE REVIEW

**Comment by cjthompson** [comment]
@reallybeard - one change Chakra made that I wasn't sure the best way to fix was it removed the property `isTruncated` on `Text`. I used `noOfLines={1}` instead but I don't know if that's going to render the same way.

**Comment by cjthompson** [comment]
Outstanding Issue #1: `StakingTable` has type errors for the `columns` property

**Comment by cjthompson** [comment]
Outstanding Issue #2: `PageTransition` is causing an invalid hooks call error

**Comment by cjthompson** [comment]
Outstanding Issue #3: `NativeCreate` has a type error were we render either words or placeholders.
```
{revealed ? words : placeholders}
```

File: src/components/Layout/Header/Header.tsx [hunk 1]
```diff
@@ -38,7 +38,7 @@ export const Header = () => {
    * Open the hidden flags menu via keypress
    */
   const handleKeyPress = useCallback(
-    event => {
+    (event: { altKey: any; shiftKey: any; keyCode: number }) => {
```
Comment (mrnerdhair) [inline src/components/Layout/Header/Header.tsx:41]:
```suggestion
    (event: { altKey: unknown; shiftKey: unknown; keyCode: number }) => {
```

File: src/components/Approval/Approval.tsx [hunk 2]
```diff
@@ -121,24 +121,21 @@ export const Approval = () => {
...
-                boxSize='60px'
-                fallback={<SkeletonCircle boxSize='60px' />}
-              />
+              {() => (
+                <Image
+                  src={quote?.sellAsset?.icon}
+                  boxSize='60px'
+                  fallback={<SkeletonCircle boxSize='60px' />}
+                />
+              )}
```
Comment (mrnerdhair) [inline src/components/Approval/Approval.tsx:138]: Just for my general edification, why is this needed?

**Review by mrnerdhair — COMMENTED** [review]
FWIW, everything LGTM so far

File: src/components/Approval/Approval.tsx (hunk 2 above)
Comment (cjthompson) [reply src/components/Approval/Approval.tsx:138]: The component change to require a function rather than a component.

**Comment by 0xdef1cafe** [comment]
more progress - all lint/type/test/build errors are fixed. 

the app currently won't load in dev or production mode with two errors to deal with

```
Error [Kind: Error] PluginManager: Duplicate pluginId: bitcoinChainAdapter 
```

```
Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
```

**Comment by 0xdef1cafe** [comment]
```
➜  web git:(react-18) ✗ yarn list react
yarn list v1.22.17
warning Resolution field "@shapeshiftoss/hdwallet-core@1.23.0" is incompatible with requested version "@shapeshiftoss/hdwallet-core@1.21.1"
warning Filtering by arguments is deprecated. Please use the pattern option instead.
├─ @snek-at/storybook-addon-chakra-ui@1.0.0
│  └─ react@17.0.2
├─ react-json-csv@2.0.0
│  └─ react@16.14.0
└─ react@18.2.0
✨  Done in 0.55s.
```

we need to replace `react-json-csv` with another library as it depends on react 16, bringing in two versions, shouldn't be hard

**Comment by 0xdef1cafe** [comment]
removing both the above dependencies gets the app to load

* [ ] fox page is broken
* [ ] colors are weird
* [ ] transaction csv download needs reimplementing

**Comment by 0xdef1cafe** [comment]
image

i broke this too, while unbreaking the app

**Comment by 0xApotheosis** [comment]
@0xdef1cafe we are getting close. Colors fixed.

We need to take a look at this:

Screen Shot 2022-07-07 at 8 12 58 pm

**Comment by gomesalexandre** [comment]
> image
> 
> i broke this too, while unbreaking the app

Fixed in 361af7a5748ba1703782367b04712f0e4d202231

**Comment by gomesalexandre** [comment]
> @0xdef1cafe we are getting close. Colors fixed.
> 
> We need to take a look at this:
> 
> Screen Shot 2022-07-07 at 8 12 58 pm

This isn't a bug and won't happen in prod: https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects

If you try to remove the `<React.StrictMode />` wrapper in: https://github.com/shapeshift/web/blob/d99f81859a0957ba7784e82b659606ab1d2c5459/src/index.tsx#L14
This goes away

tl;dr `<PluginProvider />` gets invoked twice (as in, two different instances of that context provider) with strict mode enabled. Not pretty while developing, but given how the plugins manager and the plugins context provider are explicitly decoupled, we cannot silence these.

**Comment by gomesalexandre** [comment]
One smallish regression in account/asset pages ghost buttons:

<img width="180" alt="image" src="https://user-images.githubusercontent.com/17035424/177887642-df966134

**Comment by 0xApotheosis** [comment]
> One smallish regression in account/asset pages ghost buttons:

Fixed in https://github.com/shapeshift/web/pull/2065/commits/8b942c4dd998d2db8cad3eb89d1c356eda7146dd.

How it ever worked to begin with I have no idea, the values were all wrong.

**Comment by 0xApotheosis** [comment]
Our custom [radio button implementation](https://github.com/shapeshift/web/blob/52b4559148f6ea754cbc352c34eddced0362146e/src/components/Radio/Radio.tsx) is broken with respect to showing the active radio button.

If the fix isn't obvious we might want to re-implement as per Chakra's V2 guide: https://chakra-ui.com/docs/components/radio#custom-radio-buttons

**Comment by gomesalexandre** [comment]
> Our custom [radio button implementation](https://github.com/shapeshift/web/blob/52b4559148f6ea754cbc352c34eddced0362146e/src/components/Radio/Radio.tsx) is broken with respect to showing the active radio button.
> 
> If the fix isn't obvious we might want to re-implement as per Chakra's V2 guide: https://chakra-ui.com/docs/components/radio#custom-radio-buttons

If we're talking about the `<TxFeeRadioGroup />` component, fixed in 0ef673bc4a5268e2ccddcd87f42ba5e08f82668c by using the right variant. `ghost-filled` was both semantically and visually wrong

**Comment by 0xApotheosis** [comment]
> > Our custom [radio button implementation](https://github.com/shapeshift/web/blob/52b4559148f6ea754cbc352c34eddced0362146e/src/components/Radio/Radio.tsx) is broken with respect to showing the active radio button.
> > If the fix isn't obvious we might want to re-implement as per Chakra's V2 guide: https://chakra-ui.com/docs/components/radio#custom-radio-buttons
> 
> If we're talking about the `<TxFeeRadioGroup />` component, fixed in [0ef673b](https://github.com/shapeshift/web/commit/0ef673bc4a5268e2ccddcd87f42ba5e08f82668c) by using the right variant. `ghost-filled` was both semantically and visually wrong

Excellent catch @gomesalexandre! I suspect this upgrade has made a lot of things that weren't working actually work as specified, which is now surfacing bugs with the original implementation.

**Comment by 0xApotheosis** [comment]
Storybook is fixed, and now builds correctly.

The radio buttons are still broken - I suspect there is some kind of re-render resetting the state upon click.

**Comment by 0xApotheosis** [comment]
I've fixed a regression with the `react-qr-reader` by not lazy loading it, but the stying is borked. 

TODO.

**Comment by gomesalexandre** [comment]
> I've fixed a regression with the `react-qr-reader` by not lazy loading it, but the stying is borked.
> 
> TODO.

Updated the consumption of this library to match the new API. Looks like the whole lib is currently having some error on scan with the latest published (beta) version: https://github.com/react-qr-reader/react-qr-reader/issues/239

We will have to patch it / go with another library to be able to unrug this

**Comment by 0xApotheosis** [comment]
Also rugged by the infinite loop mentioned in https://github.com/react-qr-reader/react-qr-reader/issues/266

**Comment by gomesalexandre** [comment]
> Also rugged by the infinite loop mentioned in [react-qr-reader/react-qr-reader#266](https://github.com/react-qr-reader/react-qr-reader/issues/266)

QR reader fixed in b7a159e98f31ac7702fb34e6e1d9b93dec9f67ec (`react-qr-scanner` removal) and 880ddcbdd3cc7dbd683ff94377708a0d6ab9bfa5 (implementation) using `html5-qr-reader`, which seems to be the most maintained barebones implementation of a QR reader, library/framework agnostic: without react bindings, we won't be rugged by react version bumps anymore.

Made a custom React implementation layering on top of this reader and reusing our `<QRCodeScanner />` component.

## Camera permission not granted

image

## Permission - native popup

image

## Active camera 

image

Mobile is using the front camera (tested on iPhone) which is probably not what we want, looking at it now

**Comment by gomesalexandre** [comment]
> Mobile is using the front camera (tested on iPhone) which is probably not what we want, looking at it now

Fixed in 52e51049ad1d90e3605c8c05a85d5b95d0a4b34b, we're 🥖 on mobile:

## Permission - native popup

image

## Active back camera

image

**Comment by gomesalexandre** [comment]
Removed some deprecated chakra properties in  15af528 - these were added with the Pendo plugin feature but are now invalid in chakra v2, see commit body

**Comment by gomesalexandre** [comment]
Updated react-hook-form from 7.32.2 to 7.33.1 in 2900547, which brings support from react 18 starting from v7.33.

Also 
  - added the right typings to the `useForm()` calls in the `src/context/WalletProvider/NativeWallet/` views which are currently not typed with any generic and fail type-check
  - used the `Boolean()` constructor everywhere we use error.someField as a truthy value
  
Note that these fixes are not related to the react-hook-form version bump, the bump was simply done to honor the react 18 update and make sure we have the latest types.

**Comment by 0xApotheosis** [comment]
We're getting close! The only thing I'm noticing now is that the `SideNav` "flashes" when changing between menu items.

I'm guessing there was a breaking change in one of our animation libraries, but it's not immediately clear to me what. I've pinged @reallybeard for his thoughts.

**Review by 0xdef1cafe — APPROVED** [review]

**Review by 0xApotheosis — APPROVED** [review]
🥓
